rebuild/src/functions.h

```diff
@@ -0,0 +1,164 @@
+#include "addon.h"
+#include "blst.hpp"
+#include "napi.h"
+
+/**
+ * @note The unique_ptr is used to hold the blst::P* object if a Uint8Array is
+ *       being converted instead of a PublicKey or Signature object.
+ */
+template <typename T>
+struct PointerGroup {
+    std::unique_ptr<T> unique_ptr = std::make_unique<T>();
+    T *raw_pointer = nullptr;
+};
+
+/**
+ *  Unwraps a Napi::Value to a blst::P* object. If the value is a Uint8Array it
+ *  will be converted to a blst object. If the value is a PublicKey or
+ *  Signature the pointer will be copied from the object and reused to avoid
+ *  duplication.
+ *
+ * @param[out] ptr_group PointerGroup<T> &
+ * @param[in] env const Napi::Env &
+ * @param[in] module const BlstTsAddon *
+ * @param[in] value const Napi::Value &
+ * @param[in] js_class_name const std::string &
+ * @param[in] coord_type CoordType
+ *
+ * @note This can potentially segfault if incorrect CoordType is passed. There
+ *       is a C-style cast to get this to compile as the raw pointer is set from
+ *       two different types depending on which branch of the conditional is
+ *       taken by the code. This is safe as long as the CoordType is correct for
+ *       the type of PointerGroup that is used.
+ *
+ * @note This can potentially throw a BLST_ERROR. Must be in try/catch but is
+ *       not in this function so a loop counter from the calling context can be
+ *       used in the error message
+ */
+template <typename T>
+bool unwrap_point_arg(
+    PointerGroup<T> &ptr_group,
+    const Napi::Env &env,
+    const BlstTsAddon *module,
+    const Napi::Value &value,
+    const std::string &js_class_name,
+    CoordType coord_type) {
```

Comment:
take the lengths compressed/uncompressed as input as well as the point class P1/P2 as input , then the actual code could be shortened and not get into if else

Reply:
Updated.  Converted back to a macro per our discord discussion